### ISSUE TITLE
PoC: Exits and moving around

### ISSUE THREAD

**Issue by vilmibm** [vilmibm/tildemush#38]
Once #37 is done, implement a PoC of how to move between rooms. This will likely look like sending an action to a specific exit object, like a door.

the big open question is if we'll have cardinal shortcuts; what does `north` mean when exits are just arbitrary bridges to other places? should cardinal directions be hardcoded into moving around somehow?

**Comment by modgethanc** [comment]
i wonder if `up down left right` might be better than cardinal directions, at least (maybe tied to arrow keys/nethack-like bindings as an option); i think it's hard to get too far away from the 'pick a specific direction to go to' model, and while i have a little bit of nostalgia for cardinal directions as a classic mu*/IF movement control, we don't necessarily have to be tied to it if WASD/arrow keys is a more accessible/intuitive one.

like if there was a keybinding specifically for movement (ie ctrl-UP), and a little minimap overlay to guide you?

**Comment by modgethanc** [comment]
i don't think it's too ambitious to have stacks of planes (hell, why stop there, let's have an n-ary world??); i'm thinking like how nethack has z-axis up/down controls when there are stairs present? i don't think it's that much more burdensome to model, though maybe it'll get trickier with minimap display.

**Comment by vilmibm** [comment]
haha in truth i want n-ary worlds! which is why completely free form exits appeal to me. but i guess there is no difference between a hardcoded `up, down, left, right, z-up, z-down` and then `n` arbitrary exits?

**Comment by modgethanc** [comment]
yeah, i guess if there are going to be `n` arbitrary exits, then you're really going to have to call movement by picking a specific exit, like `go treehouse_1` `go path_A`, but then that means rooms/exits are going to need identifiers that are unique in that room at least.

i think i'm still leaning towards like, if i'm in a particular room, and i remember sequence `u u l zup l zup u u r` as path to a different room, once i have that in muscle memory i should just be able to zip there, instead of being in each room and typing out `go hallway, go stairs, go window, go garden, go path_a, go path_b` etc.

is there a viable middle ground, like there's `up, down, left, right, alt-up, alt-down` for the standard world, but then have special exits that enter a more n-ary space that you can reference by name? `go portal` or something?

**Comment by vilmibm** [comment]
> is there a viable middle ground, like there's up, down, left, right, alt-up, alt-down for the standard world, but then have special exits that enter a more n-ary space that you can reference by name? go portal or something? 

this is exactly what i'm thinking. and i love your point about muscle memory travel.

>  i'm now imagining like a grand central station with 92635783 different exits that all branch into their own worlds 8)

this is the kind of magical chaos i'm here for

so here's what i'm going to do (tackling this issue next):

add support for these commands:

- [ ] `/up`, `/down`, `/left`, `/right`, `/zup`, `/zdown`.
- [ ] add `/go` (and alias `/g`) that can take the above directions
- [ ] add ability for `/go` to take an exit name
- [ ] let game objects define a `go` handler that tells the game world to move a character somewhere
- [ ] either ctrl+arrow or arrow to send a `/go` command. the opposite would interact with the prompt a la `readline`.

as for a map, i'm thinking something abstract like:

```
                   Dining Hall
                          ^   Above: Balcony
Coat Room < Gallery > Bathroom
                          v   Below: Basement
                       Foyer
```

**Comment by modgethanc** [comment]
i think that command structure is solid!

while i have maps/minimaps on the brain, i was thinking it might be nice to have a bigmap available that the user can scroll around in to plan travel, but also only have it show rooms the user has actually been in (maybe have some visual suggestion of unexplored exits to remind them of places they haven't fully explored).

in general, i think there's a balance to be struck between making navigation/movement too burdensome for most people (especially if they don't have a lot of experience navigating text-based worlds like this one), but also not take away the actual feeling of moving around in the world, if that makes sense? at least for me, a lot of the mu* immersion is that moving through rooms feels meaningful! but also portals/teleport actions are cool, too (i'd totally want to script myself a hearthstone that i could anchor in a room and pop myself there whenever i wanted to, for example)

**Comment by vilmibm** [comment]
i like the big map idea and about having it be unexplored!

about maintaining a healthy tension between intuitive moving through the world and willy nilly teleportation, i think having a hardcoded set of directional exits with the ability to put in "hearthstones" or "portkeys" for non-intuitive movement is :+1:

**Comment by vilmibm** [comment]
Marking this as closed since the basics of moving/teleporting/exits now exist in the code. they're far from perfect but other tickets can describe improvements.
